xbmc/platform/win32/filesystem/Win32WSDiscovery.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ *  Copyright (C) 2005-2021 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "Win32WSDiscovery.h"
+
+
+HRESULT CreateClientNotificationSink(CClientNotificationSink** sink)
+{
+  CClientNotificationSink* tempSink = nullptr;
+
+  if (!sink)
+    return E_POINTER;
+
+  tempSink = new CClientNotificationSink();
+
+  if (!tempSink)
+    return E_OUTOFMEMORY;
+
+  *sink = tempSink;
+  tempSink = nullptr;
+
+  return S_OK;
+}
+
+CClientNotificationSink::CClientNotificationSink() : m_cRef(1)
+{
+  m_finished = false;
+  m_serversIps.clear();
+}
+
+CClientNotificationSink::~CClientNotificationSink()
+{
+}
+
+HRESULT STDMETHODCALLTYPE CClientNotificationSink::Add(IWSDiscoveredService* service)
+{
+
+  if (!service)
+    return E_INVALIDARG;
+
+  WSD_NAME_LIST* list = nullptr;
+  service->GetTypes(&list);
+
+  LPCWSTR address = nullptr;
+  service->GetRemoteTransportAddress(&address);
+
+  if (list && address)
+  {
+    std::wstring name(list->Next->Element->LocalName);
+
+    if (name == L"Computer")
+    {
+      std::wstring addr(address);
+      m_serversIps.push_back(addr.substr(0, addr.find(L":", 0)));
+    }
+  }
+
+  return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CClientNotificationSink::Remove(IWSDiscoveredService* service)
+{
+  return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CClientNotificationSink::SearchFailed(HRESULT hr, LPCWSTR tag)
+{
+  m_finished = true;
+
+  return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CClientNotificationSink::SearchComplete(LPCWSTR tag)
+{
+  m_finished = true;
+
+  return S_OK;
+}
+
+void CClientNotificationSink::WaitSearchComplete()
+{
+  while (!m_finished)
+    Sleep(100);
+}
+
+HRESULT STDMETHODCALLTYPE CClientNotificationSink::QueryInterface(REFIID riid, void** object)
+{
+  HRESULT hr = S_OK;
+
+  if (!object)
+    hr = E_POINTER;
+
+  if (hr == S_OK)
+  {
+    *object = nullptr;
+
+    if (__uuidof(IWSDiscoveryProviderNotify) == riid)
+      *object = static_cast<IWSDiscoveryProviderNotify*>(this);
+    else if (__uuidof(IUnknown) == riid)
+      *object = static_cast<IUnknown*>(this);
+    else
+      hr = E_NOINTERFACE;
+  }
+
+  if (hr == S_OK)
+    ((LPUNKNOWN)*object)->AddRef();
+
+  return hr;
+}
+
+ULONG STDMETHODCALLTYPE CClientNotificationSink::AddRef()
+{
+  ULONG newRefCount = (ULONG)InterlockedIncrement((LONG*)&m_cRef);
+
+  return newRefCount;
```

Comment:
Doesnt look like the return for this, or release are used at all. Any need to bother with a return value?

Reply:
The return value is not used, however can not be removed because is a derived class from `IUnknown::AddRef` and needs to be same methods and params:
https://docs.microsoft.com/en-us/windows/win32/api/unknwn/nf-unknwn-iunknown-addref

Eve so, I have simplified some things in this part of the code:

```c++
HRESULT STDMETHODCALLTYPE CClientNotificationSink::QueryInterface(REFIID riid, void** object)
{
  if (!object)
    return E_POINTER;

  *object = nullptr;

  if (__uuidof(IWSDiscoveryProviderNotify) == riid)
    *object = static_cast<IWSDiscoveryProviderNotify*>(this);
  else if (__uuidof(IUnknown) == riid)
    *object = static_cast<IUnknown*>(this);
  else
    return E_NOINTERFACE;

  ((LPUNKNOWN)*object)->AddRef();

  return S_OK;
}

ULONG STDMETHODCALLTYPE CClientNotificationSink::AddRef()
{
  ULONG newRefCount = InterlockedIncrement(&m_cRef);

  return newRefCount;
}

ULONG STDMETHODCALLTYPE CClientNotificationSink::Release()
{
  ULONG newRefCount = InterlockedDecrement(&m_cRef);

  if (!newRefCount)
    delete this;

  return newRefCount;
}
```